### ISSUE TITLE
Detail.cpp Code

### ISSUE THREAD

**Issue by fabiopuddu77** [usnistgov/AGA8#26]
Hello,

I am trying to translate your code, I am talking about this part of code, did you recalculate the n0i terms?

n0i[1][3] = 4.00088; 
n0i[2][3] = 3.50031; 
n0i[3][3] = 3.50002; 
n0i[4][3] = 4.00263; 
n0i[5][3] = 4.02939; 
n0i[6][3] = 4.06714; 
n0i[7][3] = 4.33944; 
n0i[8][3] = 4;       
n0i[9][3] = 4;       
n0i[10][3] = 4;      
n0i[11][3] = 4;      
n0i[12][3] = 4;      
n0i[13][3] = 4;      
n0i[14][3] = 4;      
n0i[15][3] = 2.47906;
n0i[16][3] = 3.50146;
n0i[17][3] = 3.50055;
n0i[18][3] = 4.00392;
n0i[19][3] = 4;      
n0i[20][3] = 2.5;    
n0i[21][3] = 2.5;   

    // Ideal gas terms
    d0 = 101.325/RDetail/298.15;
    for (int i=1; i <= MaxFlds; ++i){
      n0i[i][3] = n0i[i][3] - 1;
      n0i[i][1] = n0i[i][1] - log(d0);
    }

**Comment by ianhbell** [comment]
I don't follow your question

**Comment by EricLemmon** [comment]
If I understand your question correctly, you are asking about the enthalpy and entropy reference states and that they are different from that reported in the GERG model.  If so, then you are correct that they were updated.  If you look into the code you will find  this, which shows how I did it:

```
  'Code to produce nearly exact values for n0(1) and n0(2)
  'This is not called in the current code, but included below to show how the values were calculated.  The Exit Sub above can be removed to call this code.
  T0 = 298.15
  d0 = 101.325 / RGERG / T0
  For i = 1 To MaxFlds
    n1 = 0: n2 = 0
    If th0i(i, 4) > Epsilon Then n2 = n2 - n0i(i, 4) * th0i(i, 4) / Tanh(th0i(i, 4) / T0): n1 = n1 - n0i(i, 4) * Log(Sinh(th0i(i, 4) / T0))
    If th0i(i, 5) > Epsilon Then n2 = n2 + n0i(i, 5) * th0i(i, 5) * Tanh(th0i(i, 5) / T0): n1 = n1 + n0i(i, 5) * Log(Cosh(th0i(i, 5) / T0))
    If th0i(i, 6) > Epsilon Then n2 = n2 - n0i(i, 6) * th0i(i, 6) / Tanh(th0i(i, 6) / T0): n1 = n1 - n0i(i, 6) * Log(Sinh(th0i(i, 6) / T0))
    If th0i(i, 7) > Epsilon Then n2 = n2 + n0i(i, 7) * th0i(i, 7) * Tanh(th0i(i, 7) / T0): n1 = n1 + n0i(i, 7) * Log(Cosh(th0i(i, 7) / T0))
    n0i(i, 3) = n0i(i, 3) - 1
    n0i(i, 1) = n1 - n2 / T0 + n0i(i, 3) * (1 + Log(T0))
    n0i(i, 2) = n2 - n0i(i, 3) * T0
    For j = 1 To 7: n0i(i, j) = Rsr * n0i(i, j): Next
    n0i(i, 2) = n0i(i, 2) - T0
    n0i(i, 1) = n0i(i, 1) - Log(d0)
  Next

**Comment by fabiopuddu77** [comment]
Yes I meant that.

And what about the Density that you use in the Density Detail function, which D are you using in the if (D < -epsilon) 
Because you didn't declare it before. As I have just pressure and temperature at starting point.

    ierr = 0;
    herr = "";
    if (std::abs(P) < epsilon){ D = 0; return; }
	
    tolr = 0.0000001;
    if (D > -epsilon) {
        D = P / RDetail / T;  // Ideal gas estimate
    }
    else{
        D = std::abs(D);      // If D<0, then use as initial estimate
    }

**Comment by EricLemmon** [comment]
This bit of code is only for use when the algorithm returns a vapor state when it should be liquid.  You can send it a negative number that is about three times larger (on an absolute basis) than the critical density and it will use that as an initial guess.  Otherwise it uses the ideal gas law to get the initial density starting point.  Of course, this means that you have to determine if your state point is in the liquid phase, but for a pure fluid that is easy if you have an ancillary equation for vapor pressure.

**Comment by fabiopuddu77** [comment]
How can I calculate the initial density in both of state point?

**Comment by ianhbell** [comment]
You need to figure that out yourself. Sometimes people use cubic EOS plus volume translation, for instance. This implementation is lower-level than that in REFPROP which handles things like initial guesses for you.

**Comment by fabiopuddu77** [comment]
As I can see in the code I can't calculate density from Pressure Temperature and composition mix because there is the function Pressure Detail inside the function Density. Or am I wrong?

**Comment by ianhbell** [comment]
You need to obtain the guess value for density on your own, and provide -density to that function.

**Comment by fabiopuddu77** [comment]
Ok I am trying to debug the code:

1. Because my final goal is to calculate the ZDetail, which function should I use? Because I implemented these functions:
xterms, alphar, density, pressure, setup.

2. In the function pressure, why you call the xterms function? Because I don't see any output that you use in the func pressure.

**Comment by ianhbell** [comment]
I recommend that you get a copy of the standard and read it over closely to understand how the different methods operate.

**Comment by fabiopuddu77** [comment]
Where can I find the copy of the standard?

**Comment by ianhbell** [comment]
You need to purchase it from a vendor of standards

**Comment by fabiopuddu77** [comment]
Is there any way to compile the script in static mode?

**Comment by ianhbell** [comment]
What do you mean?

**Comment by fabiopuddu77** [comment]
Create the executable for Gross_test_01 with static linking of all the libraries. That includes the libraries in the executable. Because I think now it is dynamic. But I don't know if it possible

**Comment by fabiopuddu77** [comment]
Yes, it works from a ubuntu machine, but when I try to run it from the shell docker it doesn't find the library. So I thought it was this kind of problem.
![image](https://github.com/usnistgov/AGA8/assets/61499350/c1001fc6-9848-47cc-9af4-b5c8b8329bc2)

**Comment by ianhbell** [comment]
Please show the output of your build from the first cmake call for a clean build

**Comment by fabiopuddu77** [comment]
![image](https://github.com/usnistgov/AGA8/assets/61499350/a7d024a6-c202-4b36-8cae-e4e9ae21beeb)

This is the output from ubuntu. From here I don't have any problem to run the code. But when I try to run the executable that I uploaded in the container I get the error in my previous message.

**Comment by ianhbell** [comment]
You have to build in the container too. The standard libraries differ between container and host.

**Comment by fabiopuddu77** [comment]
I can't do that because cmake and gcc doesn't exist in the machine I think and I don't know how to add it.

**Comment by ianhbell** [comment]
No choice, you have to have to build in the container. Add them to your dockerfile for the container.

Here is an example: https://github.com/usnistgov/REFPROP-tests/tree/master/docker/test
